### DISCUSSION TITLE
Python web playground interrupts BLE connection with the robot

### DISCUSSION THREAD

**Discussion by gmptrr98** [iRobotEducation/create3_docs#179] [Questions]
Hello, I'm having some troubles using Python web playground, namely connection doesn't last long, it's hard to find the robot again once connection is lost and that can lead to dangerous situation (ex. the robot keeps moving forward because it has lost connection during a test). This makes coding test very tedious. 
I tried using both a Linux 20.04 machine and a Mac, both using chrome. Is there something wrong with my Create 3 model that I just go? Should I try with another browser? Is there any other quick way to test a python code on the real Create 3?
Thank you for the help

**Comment by miniBloq** [comment]
Hi, in order to understand better what may be happening with your setup, it could be useful to download the iRobot Coding app to a mobile device (either Android or iOS based), connect, and run a program from there (all programming levels can drive the Create 3 robot). Please let us know if the connection is stable or unstable with that setup.

**Comment by gmptrr98** [comment]
It seems to be working better on the phone. Is there anyway to import my python code on the phone or do I have to build it from scratch? Why isn`t it working with my Macbook pro 2015 for more than 15/20 second?
Thank you for the help and quick response

**Reply by miniBloq** [reply]
Hi,
- Regarding that it is working better in your phone, please let us know: did you experience any interruption at all with it?
- Could you please let us know the OS version from your Mac Pro 2015. We did not experience this problem before, but being the computer 7 years old, probably there is an issue with the max supported OS version. Some older systems may not be well equipped for the Web Bluetooth API that the browser provides to our Python environment.
- iRobot Coding does not support the migration of Python code from the Python Web Playground. However, if you have an Android phone or tablet, there is a chance that the Python Web Playground may work when using Chrome, but we do **not** ensure compatibility of it with mobile devices at all.

**Comment by gmptrr98** [comment]
I experienced no interrupts with my phone.
The mac is dual booted, it has macOs bigSur 11.4 and Ubuntu 20.04 (which gives me an even more unstable connection). Bluetooth works fine with other devices but this is the first time I'm using Web bluetooth API. 
Unluckily I do not own an android device. 
How can I test my code? Is it possible to create a ROS node and launch my code using wifi connection and ROS commands?

**Reply by miniBloq** [reply]
Oh, I see, I thought that you were using a Mac computer, AND and Linux computer. So most likely this is a problem due to bad support for the browser's Web Bluetooth API in the 7-year-old hardware. Do you have any other laptop with builtin BLE that you can use?

**Reply by alsora** [reply]
Hi @gmptrr98!
Yes, you can definitely create an application that communicates with the robot via ROS 2 APIs.

Make sure that you followed the ROS 2 and WiFi setup in the Create 3 docs.
Then I suggest to start with basic ROS 2 commands to make sure that your setup is correct and you can indeed communicate.
For example, try using ROS 2 command line tools to introspect the robot and make sure that you can see all the robot's ROS entities as shown here: https://iroboteducation.github.io/create3_docs/api/ros2/

Then, you can write a Python or C++ ROS 2 application that implements your task.
I recommend to have a look at basic ROS 2 Python examples https://github.com/ros2/examples/tree/rolling/rclpy/topics and our Create 3 Python examples https://github.com/iRobotEducation/create3_examples/tree/galactic/create3_examples_py

**Reply by gmptrr98** [reply]
Perfect! Does the syntax between a python code written on the python.irobot and if I plan to use with ROS2? I mean, other then setting up proper relations between the node (subscribers and publisher) will the commands be the same (so using events, async def, await, etc..)?

**Comment by gmptrr98** [comment]
Not my self, but I could lend it from somebody else. But I can build the node to test the code with wifi right? I`m not an expert with ROS but that should be doable, also using wifi I would have a better range with it, is it correct?

**Comment by shamlian** [comment]
If you're still interested to use the Bluetooth connection with Python offline, we have made the SDK at the core of the Python Web Playground publicly available in BETA: https://github.com/iRobotEducation/irobot-edu-python-sdk/
